// ng-lock-system/src/lock.component.base.js
import { LockManagerService } from "./lock-manager.service";
import { UtilService } from "./util.service";
import { RootServiceLocator } from "./root.injector.const";
var NgLockComponentBase = /** @class */ (function () {
    function NgLockComponentBase() {
        this.util = null;
        this.isLocked$ = null;
        this.isLocked = false;
        this.lockManager = null;
        this._componentId = "";
        this._lockSubscription = null;
        this._subscriptions = [];
        this.util = RootServiceLocator.injector.get(UtilService);
        this.lockManager = RootServiceLocator.injector.get(LockManagerService);
        this.setComponentId(this.util.getId());
        this.setNgOnDestroyHook();
    }
    NgLockComponentBase.prototype.setComponentId = function (componentId) {
        this._componentId = componentId;
        this.isLocked$ = this.lockManager.listenTo(this._componentId).debounceTime(0);
        if (this._lockSubscription) {
            this.unWrapLockSubscription();
        }
    };
    NgLockComponentBase.prototype.unWrapLockSubscription = function () {
        var _this = this;
        if (this._lockSubscription) {
            this.lockManager.unListen(this._lockSubscription);
        }
        if (!this.isLocked$ || !this.isLocked$.subscribe) {
            return;
        }
        this._lockSubscription = this.isLocked$.subscribe(function (isLocked) { return _this.isLocked = isLocked; });
    };
    NgLockComponentBase.prototype.busyManager = function (value) {
        if (value === void 0) { value = true; }
        this.busy(value, value);
    };
    NgLockComponentBase.prototype.busy = function (value, managerBusy) {
        if (value === void 0) { value = true; }
        if (managerBusy === void 0) { managerBusy = false; }
        value
            ? this.lockManager.lockComponent(this._componentId, managerBusy)
            : this.lockManager.unlockComponent(this._componentId);
    };
    NgLockComponentBase.prototype.addSubscription = function (s) {
        this._subscriptions.push(s);
    };
    NgLockComponentBase.prototype.cleanupData = function () {
        this.lockManager.unListen(this._lockSubscription);
        this.lockManager.removeComponent(this._componentId);
        this.lockManager.unListen(this._subscriptions);
    };
    NgLockComponentBase.prototype.setNgOnDestroyHook = function () {
        var _this = this;
        var old = function () {
        };
        if (this.util.isFunction(this["ngOnDestroy"])) {
            old = this["ngOnDestroy"].bind(this);
        }
        this["ngOnDestroy"] = function () {
            old();
            _this.cleanupData();
        };
    };
    return NgLockComponentBase;
}());
export { NgLockComponentBase };
